### ISSUE TITLE
idea: do async IO using kernel IO polling instead of worker threads

### ISSUE THREAD

**Issue by minecrawler** [amethyst/amethyst#499]
For example the asset loader [uses worker threads](https://github.com/amethyst/amethyst/blob/70bfece81c2b290cfd43f08a0bd175e8f6427d39/amethyst_assets/src/loader.rs#L138) to load assets, however those threads could be used to do actual calculations and work. Loading assets from disk can be handed to the kernel using epoll & co. Imho, the gain would be especially high for big assets and great number of assets (which probably is normal for complex scenes), not slowing down rendering and logic while loading.

For example, Tokio and NodeJS already use that technique successfully to gain great boosts for networking and file IO loads.

What's your opinion?

**Comment by torkleyy** [comment]
If it's still possible to keep out current API, I'm all for it.

**Comment by minecrawler** [comment]
instead of threads doing the work, we would have to put a task in place, which from time to time does the polling to update the asset status and make them available to rendering.

**Comment by minecrawler** [comment]
@omni-viral as far as I can see, there is no FS library using polling at the moment. A new library would have to be created, which provides cross-platform FS operations, which are based on kernel polling mechanisms. The most prominent mechanisms already have Rust bindings, though most of them not as stable release: [epoll](https://crates.io/crates/epoll)(Linux, Android) and [kqueue](https://crates.io/crates/kqueue)(BSD, macOS, iOS), [IOCP](https://crates.io/crates/iocp)(Windows, AIX, Solaris).

While fitting this solution would be beneficial imho, I wanted to know if there is actual interest in optimized low-level IO. Since the asset loader already works at the moment, and creating such an additional library seems like a lot of effort (maybe not so much for the initial implementation, but rather for maintenance), it sure is something to keep in mind, though maybe rather as a low-prio and not critical for v1.0 :)

**Comment by OvermindDL1** [comment]
Is there really no [`libuv`](http://libuv.org/) crate for Rust or so?  That is the primary method in C/C++ nowadays (I think Python and Go and others wrap that specific library as well) as it handles all the Async IO primitives on all platform.

Hmm, some quite Googling shows that Rust's `mio` library (built in?) is the libuv replacement in Rust?

**Comment by minecrawler** [comment]
mio seems to not be built in (available on [crates.io](https://crates.io/crates/mio)). The description reads:

> ## Features
> 
> - Event loop backed by epoll, kqueue.
> - Zero allocations at runtime
> - Non-blocking TCP, UDP and Unix domain sockets
> - High performance timer system
> - Thread safe message channel for cross thread communication
> 
> ## Non goals
> 
> The following are specifically omitted from MIO and are left to the user or higher level libraries.
> 
> - File operations
> - Thread pools / multi-threaded event loop

So we would still need to implement the FS part (though it might be better than using the low-level crates)

**Comment by Rhuagh** [comment]
We already have the `Processor` systems for polling, should be able to use those to drive the polling for this. That way the interface shouldn't change. The problem will be making this work for all sources, because we'll likely have source that aren't a filesystem.

**Comment by minecrawler** [comment]
We could put everything behind a future. Stuff which needs time to resolve would be the FS and data over the network. Stuff which would be instant is binary data (genmesh?), but also behind the future in order to keep it simple and consistent. Which other ways do you have in mind @Rhuagh ?

**Comment by Rhuagh** [comment]
We've tried futures, and they were rejected because they led to more problems than they solved.

**Comment by omni-viral** [comment]
The main problem with future was sharing of the result. They can (possibly should) be used inside asset manager however.

**Comment by Rhuagh** [comment]
Possibly, my point is that it should be up to the source to decide how to drive the reading. And honestly, I don't think the performance of the current asset loading is anywhere near bad enough to focus on this right now.

**Comment by whentze** [comment]
Hi,

this is not going to be that easy. For various reasons, sync filesystem I/O is quite a bit trickier than async network I/O, and there's usually less performance to be gained - while it's quite usual for network service to have upwards of 10.000 simultaneous TCP connections, it's not at all usual to be working with that many regular files at the same time. For tens to hundres of simultaneous file operations, using threads works fine, and glibc actually implements the "async" file operations demanded by POSIX by internally using threads.

Actual, thread-free async filesystem I/O support is quite limited on most plattforms, for example:
Linux' readiness model considers every regular file to be "ready" to read from at all times.
That is, waiting for the disk to spin to a specific sector and then write data to a buffer is not considered "blocking" in the same sense that waiting for data to arrive on a network socket. Therefore, epoll is essentially useless for async file I/O. There is another API in Linux, called "aio", that *does* provide async file I/O, but mio does not use it (in fact, mio explicitly lists "file operations" as a non-goal).

While tokio does provide an "fs" module, all the operations provided by it [are marked as "blocking"](https://docs.rs/tokio-fs/0.1.0/src/tokio_fs/file/mod.rs.html#163), so they will be spawned onto tokio's threadpool.

tl;dr: If you want this, you will probably end up having to write *a lot* of plattform-specific code, essentially building for filesystem I/O what mio represents for network I/O. I think it is not worth the effort.

**Comment by anderejd** [comment]
Imho, this would be a prime exame of premature optimization.

How about focusing on creating a more impressive example instead? A basic Tetris clone does not need async IO ;)

**Comment by jojolepro** [comment]
I do have a couple of better examples. See:
https://github.com/jojolepro/key-fall
https://github.com/jojolepro/world-digger

I'm mostly on discord if you want to discuss about the engine. The link is in the repo's readme.

**Comment by Xaeroxe** [comment]
Due to the lack of player perceived issues I've downgraded the priority of this issue.

**Comment by LucioFranco** [comment]
Coming back to this async IO for FS is currently not supported in rust very well and as @anderejd mentioned is a premature optimization. I am going to close this issue if anyone has further concerns we can reopen this. thanks.
